src/main/java/graphql/nadel/engine/NadelExecutionStrategy.java

```diff
@@ -219,28 +211,67 @@ public NadelExecutionStrategy(List<Service> services,
+     * service. This error is usually present when the field has been rejected by
+     * {@link ServiceExecutionHooks#isFieldForbidden(NormalizedQueryField, Object)}
+     *
+     * @param esi              the {@link ExecutionStepInfo} for the top level field
+     * @param transformedQuery the query for that specific top level field
+     * @return a {@link GraphQLError} if the field was rejected before, otherwise empty
+     */
+    private Optional<GraphQLError> getTopLevelFieldError(ExecutionStepInfo esi, QueryTransformationResult transformedQuery) {
+        GraphQLFieldDefinition fieldDefinition = esi.getFieldDefinition();
+        String topLevelFieldId = NodeId.getId(fieldDefinition);
+        return transformedQuery.getRemovedFieldMap()
+                .getRemovedFieldById(topLevelFieldId)
+                .map(NormalizedFieldAndError::getError);
+    }
 
-        return resultToResultNode.resultToResultNode(
-                newExecutionContext,
-                new ServiceExecutionResult(dataMap, Collections.singletonList(errorMap)),
-                ElapsedTime.newElapsedTime().build(),
-                nadelContext.getNormalizedOverallQuery()
-        );
+    /**
+     * Creates the {@link RootExecutionResultNode} for a rejected field. Generally in
+     * case the underlying service was probably never called and we want to fill the
+     * overall GraphQL response with an error for that specific top level field.
+     *
+     * @param nadelContext context for the execution
+     * @param esi          the {@link ExecutionStepInfo} for the top level field
+     * @param error        the {@link GraphQLError} to put in the overall response
+     * @return {@link RootExecutionResultNode} with the specified top level field nulled out and with the given GraphQL error
+     */
+    private RootExecutionResultNode getRejectedTopLevelFieldResult(NadelContext nadelContext, ExecutionStepInfo esi, GraphQLError error) {
+        String topLevelFieldResultKey = esi.getResultKey();
+        NormalizedQueryFromAst overallQuery = nadelContext.getNormalizedOverallQuery();
+        NormalizedQueryField topLevelField = getTopLevelField(nadelContext.getNormalizedOverallQuery(), topLevelFieldResultKey);
+        return resultToResultNode.createResultWithNullTopLevelField(overallQuery, topLevelField, singletonList(error), emptyMap());
     }
 
+    /**
+     * Util function to get a top level {@link NormalizedQueryField} out of the given query.
+     *
+     * @param query                  the query to look in
+     * @param topLevelFieldResultKey the result key of the top level field to find
+     * @return the top level field if found
+     */
+    private static NormalizedQueryField getTopLevelField(NormalizedQueryFromAst query, String topLevelFieldResultKey) {
+        NormalizedQueryField topLevelField = null;
+
+        for (NormalizedQueryField candidate : query.getTopLevelFields()) {
```

Comment:
this function got into could go into the NormalizedFieldQueryFromAst itself

Reply:
Good point, thought it wasn't an API I could change